src/server/models/listItem.ts

```diff
@@ -195,6 +204,143 @@ export async function checkListItemExists({
   return total > 0;
 }
 
+export async function getPaginationValues(  props: {
+  count: number;
+  page: number;
+  params: ListsRequestParams;
+}): Promise<PaginationResults> {
+  const { count, page, params } = props;
+  const limit = 20;
+  let allPages = 0;
+  if (count > 0 && limit > 0) {
+    allPages = Math.ceil(count / limit);
+  }
+
+  const nextPrevious = await getNextPrevious({page, allPages, params});
+  const { queryString, currentPage } = nextPrevious;
+
+  const pageItems: PaginationItem[] = await getPageItems({allPages, currentPage, queryString});
+
+  const from = await getFromCount({count, limit, currentPage});
+  const to = await getToCount({currentPage, allPages, count, limit});
+
+  return {
+    pagination: {
+      results: {
+        from,
+        to,
+        count,
+        currentPage
+      },
+      previous: {
+        text: nextPrevious.previous.page.toString(),
+        href: nextPrevious.previous.queryString
+      },
+      next: {
+        text: nextPrevious.next.page.toString(),
+        href: nextPrevious.next.queryString
+      },
+      items: pageItems
+    }
+  }
+}
+
+ async function getNextPrevious(props: {page: number; allPages: number; params: ListsRequestParams }): Promise<{
+  queryString: string;
+  currentPage: number;
+  previous: {
+    page: number;
+    queryString: string;
+  };
+  next: {
+     page: number;
+     queryString: string;
+   };
+ }> {
+  const {page, allPages, params} = props;
+  let currentPage = page === undefined ? 1 : Number(page);
+  let queryStringPrevious = "";
+  let queryStringNext = "";
+  let previousPage = -1;
+  let nextPage = -1;
+
+  let queryString = await queryStringFromParams(params, true);
+  queryString = queryString.replace("&page=" + currentPage.toString(), "");
+
+  if (currentPage > allPages) {
+    currentPage = allPages;
+  }
+
+  if (currentPage > 1) {
+    previousPage = currentPage - 1;
+    queryStringPrevious = `${listsRoutes.results}?${queryString}&page=${previousPage}`;
+  }
+  if (currentPage < allPages) {
+    nextPage = currentPage + 1;
+    queryStringNext = `${listsRoutes.results}?${queryString}&page=${nextPage}`;
+  }
+  return {
+    queryString,
+    currentPage,
+    previous: {
+      page: previousPage,
+      queryString: queryStringPrevious
+    },
+    next: {
+      page: nextPage,
+      queryString: queryStringNext
+    }
+  }
+}
+
+async function getPageItems(props: {allPages: number; currentPage: number; queryString: string;}): Promise<PaginationItem[]> {
+  const { allPages, currentPage, queryString } = props;
+  const pageItems: PaginationItem[] = [];
+
+  for (let i = 1; i <= allPages; i++) {
+    let href = "";
+    if (i >= currentPage-2 && i <= currentPage+2) {
+
+      if (i !== currentPage) {
+        href = `${listsRoutes.results}?${queryString}&page=${i}`
+      }
+      pageItems.push({
+        text: (i).toString(),
+        href
+      });
+    }
+  }
+  return pageItems;
```

Comment:
looks like this is a loop which changes an array. would map be better?

Reply:
I don't think a map will work here.  We're iterating a loop up to a count and pushing some values based on the current index into an initially empty array.  The map would probably work better if we had to loop over a list of values in an array.